faiss/impl/LocalSearchQuantizer.cpp

```diff
@@ -604,54 +606,67 @@ void LocalSearchQuantizer::icm_encode_step(
     FAISS_THROW_IF_NOT(M != 0 && K != 0);
     FAISS_THROW_IF_NOT(binaries != nullptr);
 
-    for (size_t iter = 0; iter < n_iters; iter++) {
-        // condition on the m-th subcode
-        for (size_t m = 0; m < M; m++) {
-            std::vector<float> objs(n * K);
-#pragma omp parallel for
-            for (int64_t i = 0; i < n; i++) {
-                auto u = unaries + m * n * K + i * K;
-                memcpy(objs.data() + i * K, u, sizeof(float) * K);
-            }
+#pragma omp parallel for schedule(dynamic)
+    for (size_t i = 0; i < n; i++) {
+        std::vector<float> objs(K);
 
-            // compute objective function by adding unary
-            // and binary terms together
-            for (size_t other_m = 0; other_m < M; other_m++) {
-                if (other_m == m) {
-                    continue;
+        for (size_t iter = 0; iter < n_iters; iter++) {
+            // condition on the m-th subcode
+            for (size_t m = 0; m < M; m++) {
+                // copy
+                auto u = unaries + m * n * K + i * K;
+                for (size_t code = 0; code < K; code++) {
+                    objs[code] = u[code];
                 }
 
-#pragma omp parallel for
-                for (int64_t i = 0; i < n; i++) {
+                // compute objective function by adding unary
+                // and binary terms together
+                for (size_t other_m = 0; other_m < M; other_m++) {
+                    if (other_m == m) {
+                        continue;
+                    }
+
+                    if (other_m + 1 < M) {
+                        // do a single prefetch
+                        int32_t code2 = codes[i * M + other_m + 1];
+                        // for (int32_t code = 0; code < K; code += 64) {
+                        int32_t code = 0;
+                        {
+                            size_t binary_idx = (other_m + 1) * M * K * K +
+                                    m * K * K + code2 * K + code;
+                            __builtin_prefetch(binaries + binary_idx);
+                        }
+                    }
+
                     for (int32_t code = 0; code < K; code++) {
                         int32_t code2 = codes[i * M + other_m];
-                        size_t binary_idx = m * M * K * K + other_m * K * K +
-                                code * K + code2;
-                        // binaries[m, other_m, code, code2]
-                        objs[i * K + code] += binaries[binary_idx];
+                        size_t binary_idx = other_m * M * K * K + m * K * K +
+                                code2 * K + code;
+                        // binaries[m, other_m, code, code2].
+                        // It is symmetric over (m <-> other_m)
+                        //   and (code <-> code2).
+                        // So, replace the op with
+                        //   binaries[other_m, m, code2, code].
+                        objs[code] += binaries[binary_idx];
```

Comment:
Why we need to replace the access order here?

Reply:
Because the accesses becomes sequential in for-loop for code variable, thus it allows compiler to use SIMD instructions, according to the assembler code. This access reordering allows to increase the training and encoding speed about ~5x alone, according to my benchmarks.

Reply:
Btw, the same trick can be applied to the GPU code. Please feel free to alter one, or I can do that as well.

Reply:
Cool, this is an ingenious improvement! You are very welcome to improve the LSQ GPU code, if you have time.